fuse/post_history.py

```diff
@@ -0,0 +1,310 @@
+    )
+
+    df.loc[:, "last_name"] = names[0]
+    df.loc[:, "first_name"] = names[1]
+    df.loc[:, "middle_name"] = names[2]
+    return df.drop(columns=["officer_name"])
+
+
+def extract_agency(df):
+    for col in df.columns:
+        if col.startswith("agency"):
+            df[col] = (
+                df[col]
+                .fillna("")
+                .str.lower()
+                .str.strip()
+                .str.replace(r"(\w{1}? ?\w+ ?\w+? ?\w+) (.+)", r"\1", regex=True)
+            )
+    return df
+
+
+def drop_rows_missing_agency(df):
+    return df[~((df.agency.fillna("") == ""))]
+
+
+def drop_rows_missing_names(df):
+    return df[~((df.first_name == "") & (df.last_name == ""))]
+
+
+def clean_post():
+    df = (
+        pd.read_csv(deba.data("raw/post/post_officer_history.csv"))
+        .rename(columns={"officer_sex": "sex"})
+        .pipe(clean_sexes, ["sex"])
+        .pipe(split_names)
+        .pipe(clean_names, ["first_name", "middle_name", "last_name"])
+        .pipe(
+            clean_agency_row,
+            [
+                "agency",
+                "agency_1",
+                "agency_2",
+                "agency_3",
+                "agency_4",
+                "agency_5",
+                "agency_6",
+                "agency_7",
+                "agency_8",
+            ],
+        )
+        .pipe(extract_agency)
+        .pipe(
+            names_to_title_case,
+            [
+                "agency",
+                "agency_1",
+                "agency_2",
+                "agency_3",
+                "agency_4",
+                "agency_5",
+                "agency_6",
+                "agency_7",
+                "agency_8",
+            ],
+        )
+        .pipe(
+            clean_post_agency_column,
+            [
+                "agency",
+                "agency_1",
+                "agency_2",
+                "agency_3",
+                "agency_4",
+                "agency_5",
+                "agency_6",
+                "agency_7",
+                "agency_8",
+            ],
+        )
+        .pipe(gen_uid, ["first_name", "last_name", "middle_name", "agency"])
+        .pipe(gen_uid, ["first_name", "last_name", "middle_name", "agency_1"], "uid_1")
+        .pipe(gen_uid, ["first_name", "last_name", "middle_name", "agency_2"], "uid_2")
+        .pipe(gen_uid, ["first_name", "last_name", "middle_name", "agency_3"], "uid_3")
+        .pipe(gen_uid, ["first_name", "last_name", "middle_name", "agency_4"], "uid_4")
+        .pipe(gen_uid, ["first_name", "last_name", "middle_name", "agency_5"], "uid_5")
+        .pipe(gen_uid, ["first_name", "last_name", "middle_name", "agency_6"], "uid_6")
+        .pipe(gen_uid, ["first_name", "last_name", "middle_name", "agency_7"], "uid_7")
+        .pipe(gen_uid, ["first_name", "last_name", "middle_name", "agency_8"], "uid_8")
+        .pipe(drop_rows_missing_agency)
+        .pipe(drop_rows_missing_names)
+    )
+    return df
```

Comment:
There's a cleaner way to extract and present uids. Look at this example:
```python
>>> df = pd.DataFrame([['john', 'doe', 'm', 'brusly pd', 'caddo pd'], ['jane', 'lane', 'c', 'new orleans pd', 'caddo pd']], columns=['first_name', 'last_name', 'middle_name', 'agency_1', 'agency_2'])
>>> df
  first_name last_name middle_name        agency_1  agency_2
0       john       doe           m       brusly pd  caddo pd
1       jane      lane           c  new orleans pd  caddo pd
>>> stacked_agency_sr = df[['agency_1', 'agency_2']].stack()
>>> stacked_agency_sr
0  agency_1         brusly pd
   agency_2          caddo pd
1  agency_1    new orleans pd
   agency_2          caddo pd
dtype: object
>>> stacked_agency_df = stacked_agency_sr.reset_index().iloc[:, [0, 2]]
>>> stacked_agency_df.columns = ['history_id', 'agency']
>>> stacked_agency_df
   history_id          agency
0           0       brusly pd
1           0        caddo pd
2           1  new orleans pd
3           1        caddo pd
>>> names_df = df[['first_name', 'last_name', 'middle_name']].reset_index()
>>> names_df = names_df.rename(columns={'index': 'history_id'})
>>> names_df
   history_id first_name last_name middle_name
0           0       john       doe           m
1           1       jane      lane           c
>>> stacked_agency_df.merge(names_df, on='history_id', how='outer')
   history_id          agency first_name last_name middle_name
0           0       brusly pd       john       doe           m
1           0        caddo pd       john       doe           m
2           1  new orleans pd       jane      lane           c
3           1        caddo pd       jane      lane           c
```

Then just generate uid like normal. This way you only have to deal with one `agency` and one `uid` column.

Reply:
Thank you v much.